### PR TITLE
Add `:cminx: no-doc` directive

### PR DESCRIPTION
**Pull request by JackNoordhuis**
**Is this pull request associated with an issue(s)?**
There are no issues associated with these changes. The changes are proposed as a feature addition for consideration by maintainers.

**Description**
Intended to be used with configuration options:
`include_undocumented_function: true`
`include_undocumented_macro: true`
CMake files may contain internal methods/macros that should not appear in end-user documentation, but not all end-user methods/macros contain explicit documentation (yet? – aim to ease adoption.)

The use case that inspired these changes comes from an internal CMake module where steps of a large function are broken into smaller macros. The chosen approach allows all documentation comments to remain consistent by respecting a declaration that tells CMinx it should not generate documentation for a certain construct. An alternate approach would be similar to existing configuration options. A regex could be supplied to match a prefix, conveying meaning to CMinx that documentation should not be generated for matching constructs. In contrast to jumping through external files and mentally parsing regex expressions – an explicit directive conveys meaning directly in the source code for maintainers working on the affected code.

**Summary of changes**
* Add `:cminx: no-doc` directive for skipping RST generation
  - Renames `documentation_types.DocumentationType.process()` to `documentation_types.DocumentationType.process_rst()` (now considered a step of processing)
  - Reintroduces `documentation_types.DocumentationType.process()` as non-abstract entry point for processing
  - In short: `regex_replace('src/cminx/documentation_types.py', '(?<! {4}\@abstractmethod\n)( {4}def) process(\(self, writer: (?:RSTWriter|Directive)\))', '$1 process_rst$2')`
* test/test_samples: add `:cminx: no-doc` directive test

### CODE REVIEW

File: src/cminx/documentation_types.py [hunk 1]
```diff
@@ -62,12 +62,25 @@ class DocumentationType(ABC):
...
     def process(self, writer: RSTWriter) -> None:
         """
         Processes the data stored in this documentation type,
         converting it to RST form by using the given RST writer.
 
+        :param writer: RSTWriter instance that the documentation
+                       will be added to.
+        """
+        if ':cminx: no-doc' in self.doc:
+            # self.logger.info('Skipping documentation for function {function_name} because :cminx: no-doc was specified')
```
Comment (JackNoordhuis) [inline src/cminx/documentation_types.py:74]: Needs to be removed before merging. No way to access the logger instance without changing the signature but IMO this would be helpful debug-level information.

File: src/cminx/documentation_types.py [hunk 2]
```diff
@@ -62,12 +62,25 @@ class DocumentationType(ABC):
...
+            return
+        self.process_rst(writer)
+
+    @abstractmethod
+    def process_rst(self, writer: RSTWriter) -> None:
+        """
+        Processes the data stored in this documentation type,
+        converting it to RST form by using the given RST writer.
```
Comment (JackNoordhuis) [inline src/cminx/documentation_types.py:82]: Doc comments are duplicated – some insights would be appreciated

**Comment by ryanmrichard** [comment]
@JackNoordhuis Thanks for the contribution. Unfortunately, I'm pretty busy at the moment so I probably won't have a chance to look at this until maybe Friday or early next week.

@AutonomicPerfectionist I know you've moved on to other things, but if you have a couple cycles I would love some help reviewing this. If you don't have the time, no worries.

**Comment by AutonomicPerfectionist** [comment]
@ryanmrichard yep I can take a quick look in a bit

File: src/cminx/documentation_types.py [hunk 3]
```diff
@@ -62,12 +62,25 @@ class DocumentationType(ABC):
...
-    @abstractmethod
     def process(self, writer: RSTWriter) -> None:
         """
         Processes the data stored in this documentation type,
         converting it to RST form by using the given RST writer.
 
+        :param writer: RSTWriter instance that the documentation
+                       will be added to.
+        """
+        if ':cminx: no-doc' in self.doc:
```
Comment (AutonomicPerfectionist) [inline src/cminx/documentation_types.py:73]: I would recommend against checking for literal strings within the doccomment when possible, as it is error-prone and can lead to confusing errors. For example, if someone misspelled it as `:minx: no-doc` or forgot the space as in `:cminx:no-doc`, the doccomment would be output verbatim, and would likely result in errors during Sphinx generation.

CMinx already has support for parsing special-case doccomments that contain commands meant for CMinx itself, see the `@module` doccomment command. I would recommend implementing a similar mechanism for no-doc. This way, the no-doc doccomment is converted to its own token/parse rule, which can be handled by a no-op processor class. This would also eliminate the need for separate `process()` and `process_rst()` methods, since a no-doc will never be dispatched to any processor except the no-op. An example:

```cmake
#[[[ @no-doc #]]]
function(some_func)
    # Some commands
endfunction()

```

This would then be parsed into a `no-doc-command` rule, which consists of the no-doc doccomment block and the command invocation it is attached to.

Modifying the grammar a bit to disallow any `@`-commands in the doccomment header would make a misspelling of the `no-doc` command a parse error, which could point the developer directly to the line in the CMake source code that caused the error, instead of deferring it to Sphinx generation.

File: src/cminx/documentation_types.py (hunk 1 above)
Comment (AutonomicPerfectionist) [reply src/cminx/documentation_types.py:74]: If checking for `:cminx: no-doc` is the way you decide to go forward, you could change this method to a `preprocess()` method that can be overridden by subclasses. This way, you could override it in the function/method processor to provide more detailed information, such as the name of the function that was skipped.

File: tests/test_samples/no-doc_directive.cmake [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+function(undocumented_function)
+endfunction()
+
+#[[[
+# This is an undocumented macro
+#
+# :cminx: no-doc
+#]]
+macro(undocumented_macro)
+endmacro()
```
Comment (AutonomicPerfectionist) [inline tests/test_samples/no-doc_directive.cmake:15]: Would be good to have another test case with both no-doc and regular documented functions and macros. Also, testing the directive on non-function/macro definitions would be good, as the current implementation should work on all supported documentable commands.

Additionally, the name of this test should be changed, as the `:cminx:` part is not a directive but rather a fake option to the implicit `py:function` directive that is generated.

**Review by AutonomicPerfectionist — CHANGES_REQUESTED** [review]
I believe that while this change appears functional, it could be improved by modifying the grammar to create a new `@no-doc` CMinx directive, similar to the `@module` directive used in module doccomments.

File: src/cminx/documentation_types.py (hunk 3 above)
Comment (JackNoordhuis) [reply src/cminx/documentation_types.py:73]: Yep. Swallowing all `@` commands as part of the grammar is a good suggestion. After checking the `@module` command implementation I had a similar thought, noting it was part of the grammar. This looks like the best approach to maintaining consistency and providing useful error messages.

Comment (ryanmrichard) [reply src/cminx/documentation_types.py:73]: It looks like the general consensus is that this PR should be implemented as `@no-doc` rather than as `:cminx: no-doc`. Is that correct? If so, @JackNoordhuis does it make sense to keep this PR open? If we decide to close it, I'll create an issue and try to find some time to work on adjusting the grammar.

Comment (JackNoordhuis) [reply src/cminx/documentation_types.py:73]: @ryanmrichard Yes, this PR is obsolete. I have a [branch on my fork](https://github.com/CMakePP/CMinx/compare/master...JackNoordhuis:CMinx:no-doc-docstring) that has passing tests for `@` doc-string support. I've been busy this week and got sidetracked writing benchmarks for my changes so I haven't opened a new PR yet. My grammar changes just allow the AST visitor to process doc-comments with lines containing the `@` commands separately (refactors the old module nodes to generic doc command nodes.)  I spent a few hours trying to get the ANTLR grammar to parse the doc-comments (clean the lines, find `@` commands, etc.) but settled on parsing the command names and parameters by hand in Python. There are some more in-depth changes to the processing/handling of the commands themselves (escaped spaces in parameters, escaping newlines to continue parameters, etc.) that I've made with plugin/extension support in mind (future support for different output formats?)  mostly inspired by doxygen-style commands. I should have time to clean up the code and benchmark suite to open separate PR's sometime this weekend.

Comment (ryanmrichard) [reply src/cminx/documentation_types.py:73]: @JackNoordhuis That sounds amazing. I really appreciate the interest and willingness to help add a new feature!

**Comment by ryanmrichard** [comment]
Converted this PR to issue #160
